core/src/tile/tileManager.cpp

```diff
@@ -31,114 +31,173 @@ bool TileManager::updateTileSet() {
+                    removeTile(setTilesIter);
+                    tileSetChanged = true;
+                } else {
+                    ++setTilesIter;
+                }
+            } else if (*visTilesIter < setTilesIter->first) {
+                // tileSet is missing an element present in visibleTiles (shouldn't occur)
+                ++visTilesIter;
+            } else {
+                // tiles in both sets match, move on
+                ++setTilesIter;
+                ++visTilesIter;
+            }
+        }
     }
-
+    
     return tileSetChanged;
 }
 
 void TileManager::addTile(const TileID& _tileID) {
+    
+    std::shared_ptr<MapTile> tile(new MapTile(_tileID, m_view->getMapProjection()));
+    m_tileSet[_tileID] = tile;
+    
+    //Add Proxy if corresponding proxy MapTile ready
+    updateProxyTiles(_tileID, m_view->isZoomIn());
 
-    m_tileSet[_tileID].reset(nullptr); // Emplace a unique_ptr that is null until tile finishes loading
-
-    std::future<MapTile*> incoming = std::async(std::launch::async, [&](TileID _id) {
-
-        MapTile* tile = new MapTile(_id, m_view->getMapProjection());
-
-        for (const auto& source : m_dataSources) {
-            
-            logMsg("Loading tile [%d, %d, %d]\n", _id.z, _id.x, _id.y);
+    std::future< std::shared_ptr<MapTile> > incoming = std::async(std::launch::async, [&](TileID _id) {
+        
+        // Check if tile to be loaded is still required! (either not culled from m_tileSet and not logically deleted)
+        // if not set the shared state of this async's future to "null"
+        if (m_tileSet.find(_id) != m_tileSet.end()) {
+            auto tile = m_tileSet[_id];
             
-            if ( ! source->loadTileData(*tile)) {
-                logMsg("ERROR: Loading failed for tile [%d, %d, %d]\n", _id.z, _id.x, _id.y);
-            } else {
+            // Now Start fetching new tile
+            for (const auto& dataSource : m_dataSources) {
+                
+                logMsg("Loading tile [%d, %d, %d]\n", _id.z, _id.x, _id.y);
+                if ( ! dataSource->loadTileData(*tile)) {
+                    logMsg("ERROR: Loading failed for tile [%d, %d, %d]\n", _id.z, _id.x, _id.y);
+                }
                 
-                std::shared_ptr<TileData> tileData = source->getTileData(_id);
+                auto tileData = dataSource->getTileData(_id);
                 
                 for (auto& style : m_scene->getStyles()) {
-                    style->addData(*tileData, *tile, m_view->getMapProjection());
+                    if (tileData) {
+                        style->addData(*tileData, *tile, m_view->getMapProjection());
+                    }
                 }
-                
             }
             
-            
+            return tile;
+        } else {
+            // tile was deleted (went out off view) before it could load its data
+            // remove its proxies also
+            cleanProxyTiles(_id);
```

Comment:
Calling `cleanProxyTiles` from async threads risks concurrent modification, we should fix that before merging